src/ucm/util/reloc.c

```diff
@@ -302,21 +312,103 @@ static ucs_status_t ucm_reloc_apply_patch(ucm_reloc_patch_t *patch,
     return ctx.status;
 }
 
-static void *ucm_dlopen(const char *filename, int flag)
+void *ucm_dlopen(const char *filename, int flag)
 {
+    void *handle = NULL;
     ucm_reloc_patch_t *patch;
-    void *handle;
+    void *rpath_module;
+    Dl_serinfo stub;
+    Dl_serinfo *serinfo;
+    Dl_info dl_info;
+    int res;
+    int i;
+    char rpath_path[PATH_MAX];
+    struct stat file_stat;
+
+    ucm_debug("open module: %s, flag: %x", filename, flag);
 
     if (ucm_reloc_orig_dlopen == NULL) {
-        ucm_fatal("ucm_reloc_orig_dlopen is NULL");
-        return NULL;
+        ucm_reloc_orig_dlopen = ucm_reloc_get_orig(ucm_reloc_dlopen_patch.symbol,
+                                                   ucm_reloc_dlopen_patch.value);
+
+        if (ucm_reloc_orig_dlopen == NULL) {
+            ucm_fatal("ucm_reloc_orig_dlopen is NULL");
+            return NULL;
+        }
     }
 
     handle = ucm_reloc_orig_dlopen(filename, flag);
+    if (handle != NULL) {
+        goto apply_patch;
+    }
+
+    if (!ucm_global_opts.dlopen_process_rpath) {
```

Comment:
if ucm_global_opts.dlopen_process_rpath is enabled, we should NOT try orig dlopen first. this is because the search priority may not be correct if not done according to rpath.